### PR TITLE
Generalize the underlying monad stack

### PR DESCRIPTION
**Pull request by isovector**
Addresses #240

### CODE REVIEW

File: src/Brick/Forms.hs [hunk 1]
```diff
@@ -112,7 +112,7 @@ import Lens.Micro
 --  * @b@ - the form field's internal state type
 --  * @e@ - your application's event type
 --  * @n@ - your application's resource name type
-data FormField a b e n =
+data FormField a b e n m =
```
Comment (masaeedu) [inline src/Brick/Forms.hs]: Might want to add a comment on the line above saying what `m` is.

File: brick.cabal [hunk 2]
```diff
@@ -110,350 +110,3 @@ library
                        word-wrap >= 0.2
   if impl(ghc < 8.0)
     build-depends:     semigroups
-
```
Comment (ChrisPenner) [inline brick.cabal]: Assuming stack wiped all this out by accident?

File: src/Brick/Main.hs [hunk 3]
```diff
@@ -242,19 +247,19 @@ customMainWithVty :: (Ord n)
...
                   -- ^ The application.
                   -> s
                   -- ^ The initial application state.
-                  -> IO (s, Vty)
+                  -> m (s, Vty)
 customMainWithVty initialVty buildVty mUserChan app initialAppState = do
     let run vty rs st brickChan = do
             result <- runWithVty vty brickChan mUserChan app rs st
-                `E.catch` (\(e::E.SomeException) -> shutdown vty >> E.throw e)
+                -- `E.catch` (\(e::E.SomeException) -> shutdown vty >> E.throw e)
```
Comment (kquick) [inline src/Brick/Main.hs]: This helps restore the normal vty mode on exception.   Did you have something elsewhere that was handling this instead?

Comment (isovector) [reply src/Brick/Main.hs]: I didn't, but it's trivial to switch it into a `MonadCatch`. 

I needed to generalize the underlying monad for a side project, and thought I should upstream the results. I'd be happy to make this PR less shit if the maintainers are interested --- but if it's a definite "no merge" I didn't want to waste my time.

**Comment by jtdaugherty** [comment]
Closed via https://github.com/jtdaugherty/brick/issues/240.

**Pull request closed without merging**